standards/SPIFFE_Federation.md

```diff
@@ -0,0 +1,316 @@
+Trust Domain:			"prod.example.com"
+```  
+*Figure 1: Example SPIFFE bundle endpoint configuration for trust domain prod.example.com. Administrators configure SPIFFE control planes to retrieve foreign trust bundles through a bundle endpoint configuration.*
+
+When a control plane internally distributes trust bundles to workloads, the association between trust domain name and trust bundle MUST be communicated. Please see the [Security Considerations](#7-security-considerations) section for more information about the sensitivity of these parameters.
+
+![SPIFFE bundle distribution](https://raw.githubusercontent.com/evan2645/spiffe/ecb9fb894e1a13fae821c370f7b5ddeee634d1c8/standards/img/spiffe_bundle_distribution.png)  
+*Figure 2: After retrieving a foreign SPIFFE trust bundle, control planes distribute both the trust domain name and the corresponding bundle to internal workloads. Workloads use this configuration to validate identities in the foreign trust domain. For details about the trust bundle contents, refer to The [SPIFFE Trust Domain and Bundle][3], specifically the [SPIFFE Bundle Format][6] and [SPIFFE Bundle Example][4] sections.*
+
+The requirements in this section apply to all SPIFFE bundle endpoint servers and clients. Individual SPIFFE bundle endpoint profiles MAY add further requirements.
+
+### 5.2. Endpoint Profiles
+An endpoint profile describes both the transport protocol and authentication method that should be used when serving or consuming a bundle endpoint.
+
+The following sections describe the supported bundle endpoint profiles.
+
+#### 5.2.1. Web PKI (`https_web`)
+The `https_web` profile leverages publicly trusted certificate authorities to provide a low-friction path for configuring SPIFFE federation. It behaves identically to the "https" URLs that most people are familiar with when accessing web pages in their browsers. In this profile, the bundle endpoint server uses a certificate issued by a public CA, obviating the need for additional client configuration; endpoints using the `https_web` profile type are authenticated using the same public CA certificates commonly installed in modern operating systems.
+
+Please see the [Security Considerations](#7-security-considerations) section for more information about the use of public certificate authorities.
+
+##### 5.2.1.1. Endpoint URL Requirements
+Bundle endpoint URLs utilizing `https_web` MUST have the scheme set to `https` and MUST NOT include userinfo in the authority component. Other components of the URL (as defined by [RFC 3986 section 3][7]) are not constrained by this specification.
+
+For example, the URL `https://host.example.com/trust_domain` is a valid SPIFFE bundle endpoint URL for the `https_web` profile type.
+
+##### 5.2.1.2. Endpoint Parameters
+The `https_web` profile MUST NOT require any additional parameters beyond those required for every profile (namely the trust domain name, profile type, and endpoint URL) in order to function.
```

Comment:
I think this should also include a parameter for CA certificate bundle for the endpoint. There's an implied default value that works in most cases, the system certificate pool, but it's not mandated that the certificate used for the endpoint needs to be issued by a public certificate authority(line 127 has that as a SHOULD, not a MUST).

Reply:
The canonical way to trust a private CA that works with RFC 6125 style validation (i.e. Web PKI) is to add the private root to the system store. If you're trying to accomplish CA pinning, my interpretation of this text is that it is still allowed so long as the implementation does not require it in order to work.

I think this brings up a related point though about `https_spiffe`... we should probably clarify that the parameter is optionally supplied but _not_ optionally supported.

Reply:
If it's still allowed to pin a CA, that should be good enough for me.

About adding the CA to the system store, i'm not sure that's the best approach for this. That ends up trusting the CA for anything on the machine. Even for spire-server, it means that it's less secure to have multiple federation relationships that have different private CAs. So it might still be worth pointing out the implicit parameter here and making it optionally configurable.

Reply:
When we were designing the different profiles, I had assumed that people who wanted to use private CAs and limit their scope would choose `https_spiffe`.  In that profile, the endpoint is identified by a specific machine identity and authenticated by a specific cryptographic authority.

To the extent we thought people would use private CAs, I had assumed it would be in a large enterprise environment where the enterprise maintains its own CA, and all machines get the CA cert installed at provisioning time.

The use case you're describing seems to be about pinning specific bundle endpoints to specific (non-SPIFFE) CAs, which isn't something I thought much about as we were drafting this document. @sorindumitru can you talk a little bit more about your use case in the comments.  Why do you want to do it that way instead of using `https_spiffe`?

Reply:
This is an interesting use case. And I, too, would like to hear more from @sorindumitru. Thanks for bringing this up!

So far, I think that we should certainly consider this use case, but shouldn't block this PR. Since the intent of `https_web` is to achieve a balance of security/paranoia with maximum compatibility (by "obviating the need for additional client configuration"), an additional parameter to specify a custom CA root cert may best belong in a new endpoint profile. (We'll also need to reconcile support here for a custom CA with redirect support. For example, if a custom CA parameter is specified but the endpoint returns a redirect to an endpoint which chains up to a different, perhaps well known public CA, what is a client supposed to do? And would a custom CA parameter override any use of the system-wide cert store or supplement it?)

Reply:
The use case I had in mind is for when the different trust domains can't talk to each other directly but are still reachable through various gateways/proxies so `https_spiffe` might not work. An example would be `prod.example.com` couldn't directly talk with `dev.example.com` to fetch the bundle, but `dev` could still reach `prod` through gateways.

In that case the domains could reach out to a third party HTTP service that is reachable from both places to get access to the trust bundles. This case doesn’t require multiple certificates but I still think it would be wrong to add that CA to system pool. If it’s in the system pool, the same CA would be used for all TLS connections, e.g. TLS database connections.

I’m definitely ok with not blocking the PR on this. As Evan said, the wording is `MUST NOT require` so as long as it’s an optional parameter, implementations are free to have this.

I’m not really convinced that this needs to be a different profile, it just makes the root of trust for the endpoint configurable, but I can understand that this might make it harder for domains to federate in the future if they start with a private CA.